File: mobile/android/ZTECompanion/app/src/main/java/com/ztecompanion/feature/bandlock/BandLockViewModel.kt
package com.ztecompanion.feature.bandlock

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ztecompanion.core.model.BandConfig
import com.ztecompanion.core.network.UbusClient
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import javax.inject.Inject

data class BandLockState(
    val selectedNRBands: Set<Int> = emptySet(),
    val selectedLTEBands: Set<Int> = emptySet(),
    val isLoading: Boolean = false,
    val error: String? = null,
    val successMessage: String? = null,
)

@HiltViewModel
class BandLockViewModel @Inject constructor(
    private val ubusClient: UbusClient,
) : ViewModel() {

    private val _state = MutableStateFlow(BandLockState())
    val state: StateFlow<BandLockState> = _state.asStateFlow()

    fun toggleNRBand(band: Int) {
        val current = _state.value.selectedNRBands.toMutableSet()
        if (band in current) current.remove(band) else current.add(band)
        _state.value = _state.value.copy(selectedNRBands = current, successMessage = null)
    }

    fun toggleLTEBand(band: Int) {
        val current = _state.value.selectedLTEBands.toMutableSet()
        if (band in current) current.remove(band) else current.add(band)
        _state.value = _state.value.copy(selectedLTEBands = current, successMessage = null)
    }

    fun applyNRLock() {
        val bands = _state.value.selectedNRBands.sorted().joinToString(",")
        if (bands.isBlank()) return
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null, successMessage = null)
            try {
                // Lock NSA bands
                val nsaParams = buildJsonObject {
                    put("nr5g_type", "nsa")
                    put("nr5g_band", bands)
                }
                ubusClient.call("zte_nwinfo_api", "nwinfo_set_nrbandlock", nsaParams)
                // Lock SA bands
                val saParams = buildJsonObject {
                    put("nr5g_type", "sa")
                    put("nr5g_band", bands)
                }
                ubusClient.call("zte_nwinfo_api", "nwinfo_set_nrbandlock", saParams)
                _state.value = _state.value.copy(
                    isLoading = false,
                    successMessage = "NR bands locked to: $bands",
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false, error = e.message)
            }
        }
    }

    fun applyLTELock() {
        val bands = _state.value.selectedLTEBands.sorted().joinToString(",")
        if (bands.isBlank()) return
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null, successMessage = null)
            try {
                val params = buildJsonObject {
                    put("is_lte_band", "1")
                    put("lte_band_mask", bands)
                    put("is_gw_band", "0")
                    put("gw_band_mask", "")
                }
                ubusClient.call("zte_nwinfo_api", "nwinfo_set_gwl_bandlock", params)
                _state.value = _state.value.copy(
                    isLoading = false,
                    successMessage = "LTE bands locked to: $bands",
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false, error = e.message)
            }
        }
    }

    fun unlockAll() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, error = null, successMessage = null)
            try {
                ubusClient.call("zte_nwinfo_api", "nwinfo_rest_band_rat")
                _state.value = _state.value.copy(
                    isLoading = false,
                    selectedNRBands = emptySet(),
                    selectedLTEBands = emptySet(),
                    successMessage = "All bands unlocked",
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false, error = e.message)
            }
        }
    }
}

File: mobile/ios/ZTECompanion/Features/BandLock/BandLockViewModel.swift
import SwiftUI

@Observable
@MainActor
final class BandLockViewModel {
    var config: BandConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    private let client: UbusClient
    private let authManager: AuthManager

    init(client: UbusClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func toggleNRBand(_ band: String) {
        if config.nrBands.contains(band) {
            config.nrBands.remove(band)
        } else {
            config.nrBands.insert(band)
        }
    }

    func toggleLTEBand(_ band: String) {
        if config.lteBands.contains(band) {
            config.lteBands.remove(band)
        } else {
            config.lteBands.insert(band)
        }
    }

    func applyNRLock() async {
        guard !config.nrBands.isEmpty else {
            showMessage("Select at least one NR band", isError: true)
            return
        }
        isLoading = true
        let token = authManager.sessionToken
        let bandStr = config.nrBandString

        do {
            // Lock NSA bands
            let (_, _) = try await client.call(
                sessionToken: token, object: "zte_nwinfo_api",
                method: "nwinfo_set_nrbandlock",
                params: ["nr5g_type": "nsa", "nr5g_band": bandStr]
            )
            // Lock SA bands
            let (_, _) = try await client.call(
                sessionToken: token, object: "zte_nwinfo_api",
                method: "nwinfo_set_nrbandlock",
                params: ["nr5g_type": "sa", "nr5g_band": bandStr]
            )
            showMessage("NR bands locked: \(bandStr)", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func applyLTELock() async {
        guard !config.lteBands.isEmpty else {
            showMessage("Select at least one LTE band", isError: true)
            return
        }
        isLoading = true
        let token = authManager.sessionToken

        do {
            let (_, _) = try await client.call(
                sessionToken: token, object: "zte_nwinfo_api",
                method: "nwinfo_set_gwl_bandlock",
                params: [
                    "is_lte_band": "1",
                    "lte_band_mask": config.lteBandString,
                    "is_gw_band": "0",
                    "gw_band_mask": ""
                ]
            )
            showMessage("LTE bands locked: \(config.lteBandString)", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    func unlockAll() async {
        isLoading = true
        let token = authManager.sessionToken

        do {
            let (_, _) = try await client.call(
                sessionToken: token, object: "zte_nwinfo_api",
                method: "nwinfo_rest_band_rat", params: [:]
            )
            config = .empty
            showMessage("All bands unlocked", isError: false)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }
        isLoading = false
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
